src/production/phases/CodeSplittingPhase.ts

```diff
@@ -0,0 +1,189 @@
+     *
+     * @param target
+     * @param parentId
+     */
+    traceCircularDependency: function (target: Module, parentId: number): boolean {
+      let traced = false;
+      // console.log(`traceCircularDependency ${parentId}`);
+      for (const { module } of target.moduleTree.dependants) {
+        // if we resolve to the parent or entryModuleId
+        // we're safe to assume we're contained within the split
+        if (
+          module.moduleId === entryModuleId ||
+          module.moduleId === parentId
+        ) {
+          traced = true;
+        } else {
+          traced = this.traceCircularDependency(module, parentId);
+          if (!traced) break;
+        }
+      }
+      return traced;
+    },
+    /**
+     * traceOrigin accepts a target of type Module
+     * it will validate against the already processed modules
+     * and returns true if it origins back to the possibleSplitEntry Module
+     *
+     * @param target
+     */
+    traceOrigin: function (target: Module, parentId: number): boolean {
+      const { moduleId, moduleTree: { dependants } } = target;
+      // console.log(`traceOrigin ${moduleId} ${target.getShortPath()}`);
+
+      if (moduleId in this.visited) {
+        // @todo: log circular trace to console
+        // console.log(`circular dependency detected in ${target.getShortPath()}`);
+        return this.traceCircularDependency(target, parentId);
+      }
+
+      this.visited[moduleId] = true;
+      let traced = false;
+      for (const { module } of dependants) {
+        if (module.moduleId === entryModuleId) {
+          traced = true;
+        } else if (module.moduleId in productionContext.splitEntries.ids) {
+          // the module is a dynamic module and not the entry, so false!
+          // we flagAsCommonsEligible here!
+          target.flagAsCommonsEligible();
+          traced = false;
+        } else {
+          traced = this.traceOrigin(module, moduleId);
+        }
+        if (!traced) {
+          // @todo: config to flag all shared commons?
+          // target.flagAsCommonsEligible();
+          break;
+        }
+      }
+      // console.log(`tracedOrigin ${moduleId} to: ${traced}`)
+      return traced;
+    },
+    /**
+     * traverseDependencies accepts a target of type Module
+     * it will check the modules that reference this module to validate isolation
+     * if it's an isolated module (no other references) it will be included
+     * in the splitted bundle. Also the dependencies of this module will be checked
+     * to see if we can integrate those too
+     *
+     * @param target
+     */
+    traverseDependencies: function (target: Module, entry: boolean = false): boolean {
+      const { moduleId, moduleTree: { importReferences: { references } } } = target;
+
+      // we already traversed this module, so we can skip it
+      // this happens in case of circular deps
+      if (!entry && moduleId in this.traversed) {
```

Comment:
`'key' in 'object'` vs `!!object[key]`

Reply:
`object[key]` unless your key can be `0`

Reply:
key can be 0, because we start from 0... i can make the moduleId counter start at 1?